core/src/main/java/com/linecorp/armeria/client/endpoint/FileWatcherRunnable.java

```diff
@@ -0,0 +1,144 @@
+    private static final Logger logger = LoggerFactory.getLogger(FileWatcherRunnable.class);
+
+    private final WatchService watchService;
+    private final Supplier<Collection<FileWatchEvent>> ctxSupplier;
+
+    /**
+     * Initializes a runnable which watches files.
+     * @param watchService the {@code WatchService} to poll events from
+     * @param ctxSupplier supplier which contains the callback invoked for each event
+     */
+    FileWatcherRunnable(WatchService watchService, Supplier<Collection<FileWatchEvent>> ctxSupplier) {
+        this.watchService = watchService;
+        this.ctxSupplier = ctxSupplier;
+    }
+
+    @Override
+    public void run() {
+        try {
+            WatchKey key;
+            while ((key = watchService.take()) != null) {
+                for (WatchEvent<?> event : key.pollEvents()) {
+                    if (event.kind().type() == Path.class) {
+                        @SuppressWarnings("unchecked")
+                        final Path watchedPath = ((Path) key.watchable())
+                                .resolve(((WatchEvent<Path>) event).context());
+
+                        if (event.kind().equals(ENTRY_DELETE)) {
+                            logger.warn("ignoring deleted file: {}", watchedPath);
+                            continue;
+                        }
+
+                        final Path realFilePath;
+                        try {
+                            realFilePath = watchedPath.toRealPath();
+                        } catch (IOException e) {
+                            logger.warn("skipping -- unable to get real path for {}", watchedPath, e);
+                            continue;
+                        }
+
+                        runCallback(realFilePath);
+                    } else {
+                        if (event.kind().equals(OVERFLOW)) {
+                            logger.debug("watch event may have been lost");
+                        } else {
+                            logger.debug("ignoring unexpected event type: {}", event.kind().name());
+                        }
+                    }
+                }
+
+                final boolean reset = key.reset();
+                if (!reset) {
+                    logger.warn("aborting file watch due to unexpected error");
+                    break;
+                }
+            }
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            logger.trace("file watching thread interrupted");
+        } catch (ClosedWatchServiceException e) {
+            // do nothing
+        }
+    }
+
+    private void runCallback(Path filePath) {
+        ctxSupplier.get().stream().filter(
+                ctx -> filePath.startsWith(ctx.getDirPath())).forEach(ctx -> {
+            try {
+                ctx.runCallback();
+            } catch (Exception e) {
+                logger.warn("unexpected error from listener: {} ", filePath, e);
```

Comment:
Could you capitalize the first character of all log messages in this PR? e.g. `unexpected error` -> `Unexpected error`

Reply:
done! I didn't apply capitalization to exception messages for now.. but seems like the convention is mixed within the project. would you prefer exception messages be capitalized as well?

Reply:
For exception message, please capitalize only when it's a full sentence. No need to capitalize otherwise.